src/util/sharing_map.h

```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
   iterate(n, f);
 }
 
+SHARING_MAPT(void)::add_item_if_not_shared(
+  const innert &container,
+  const innert &inner,
+  delta_viewt &delta_view,
+  const bool only_common) const
+{
+  SM_ASSERT(!container.empty());
+  SM_ASSERT(!inner.empty());
+
+  std::stack<const innert *> stack;
+  stack.push(&inner);
+
+  do
+  {
+    const innert *ip = stack.top();
+    stack.pop();
+
+    SM_ASSERT(!ip->empty());
+
+    if(ip->is_internal())
+    {
+      const to_mapt &m = ip->get_to_map();
+      SM_ASSERT(!m.empty());
+
+      for(const auto &item : m)
+      {
+        const innert *i = &item.second;
+        stack.push(i);
+      }
+    }
+    else
+    {
+      SM_ASSERT(ip->is_container());
+
+      if(ip->shares_with(container))
+        return;
+
+      const leaft &l1 = container.get_container().front();
+
+      const leaf_listt &ll = ip->get_container();
+      SM_ASSERT(!ll.empty());
+
+      for(const auto &l : ll)
```

Comment:
Comment explaining this list? It looks like you're sure `container` doesn't have multiple entries but `ip->get_container()` might do, and if it does and an entry is shared that means all further entries are too?

Reply:
Yes, `add_item_if_not_shared()` is called by `get_delta_view()` only for containers which only have a single element. If we know that the leaf in the first container has the same key as a leaf in the second container (either because they share or the keys compare equal), then since the maps cannot contain duplicate keys, we can ignore all further keys in the second container since they will be different from the key of the leaf in the first container. We can ignore the further leafs in the second container since `map1.get_delta_view(map2, delta_view, only_common=false)` does not include items that are only in `map2`.